### ISSUE TITLE
Some components do not always get injected

### ISSUE THREAD

**Issue by avm99963** [avm99963/infinitegforums#21]
As mentioned by bkc56 at [`pekb/thread/67965396?msgid=85613285`](http://go/H2hkE7), sometimes when opening a profile in a new tab, the previous posts links are not shown sometimes.

I also had this issue some weeks ago, but neither me nor bkc56 have been able to reproduce it again (it happens from time to time). However, I have located 2 weak points which might cause this and other similar problems to happen, and should be fixed:

- [X] Mutation observer at `//content_scripts/console_inject.js`: it observes changes inside the `.scrollable-content` element, which might not exist yet when the script loads.
- [x] Mutation observer at `//content_scripts/console_inject.js`: it might start observing changes late: after some changes have already been made. Check whether these changes have been made before starting observing changes, to catch those.
- [x] Mutation observer at `//content_scripts/console_inject.js`: when adding the "previous posts" link, instead of watching the `ec-user > div` element, watch the `ec-user .main-card .header > .name > span` element, which gets added later, and is the `span` containing the username, which is used to add the previous posts links (and it might not always be present if we detect the other element).
- [x] Intersection observer at `//content_scripts/console_inject.js`: it's initialized with the `.scrollable-content` element, which might not exist yet when the script loads. This is much harder to solve and requires quite a bit of research to be done. However, the only problem this can cause is to break the infinite scroll feature, so it's not related to the problem being reported here.
- [ ] The dark mode switch is added even if the `header .right-control` element doesn't exist at `//content_scripts/console_inject.js`. Maybe we could also use the mutationObserver to add it after the element is added to the page.

**Comment by avm99963** [comment]
Commit e4cac40 was meant to make a reference to this issue.

**Comment by avm99963** [comment]
I just experienced an issue similar to this one: the profile indicator didn't get injected into the page when loading a thread for the first time. I found out it was because the mutationObserver which should detect when the OP's username gets added to the page is set up after this happens, and so never detects it. I've experienced this while compiling Chromium, which uses 100% of the CPU resources.

This has been fixed by looking for the OP's username one time manually before starting the mutationObserver (this change has yet to be merged).

This might also be a possible explanation for the issue described here (so the cause might actually be the opposite of what we believed). Therefore, I'll renaming this issue to "Some components do not always get injected" (originally: "Previous posts links do not always show") to cover all components which do not get injected (for now: profile indicator and previous posts links).

TODO(@avm99963):

- [ ] Find more weak points in the code –places where starting a mutationObserver late could mean we're missing things– and add them to the list in the main message.
